### ISSUE TITLE
Set handle_collision to true causing the training process slow

### ISSUE THREAD

**Issue by delaprada** [reyuwei/NIMBLE_model#12]
Hi!

I try to use `nimble_layer` as a differentiable layer in a neural network. But when I set `handle_collision` to true, the training process of the neural network would be very slow because the calculation below are executed in CPU:

```
def mesh_collision(self, floating_verts, floating_verts_normals, steady_verts, steady_faces):
    ### go to trimesh
    batch_size = floating_verts.shape[0]
    for i in range(batch_size):
        mesh_muscle = trimesh.Trimesh(steady_verts[i].detach().cpu().numpy(),
                                    steady_faces.squeeze().detach().cpu().numpy())
        skin_in_muscle = mesh_muscle.contains(floating_verts[i].detach().cpu().numpy())
        skin_surf_in_muscle = self.skin_v_surface_mask & torch.from_numpy(skin_in_muscle).to(self.device)
        if skin_surf_in_muscle.sum() <= 1:
            continue
        inside_verts = floating_verts[i][skin_surf_in_muscle].reshape(-1, 3)
        inside_verts_normal = floating_verts_normals[i][skin_surf_in_muscle].reshape(-1, 3)

        ## moving target using ray-triangle hit
        locations, index_ray, index_tri = mesh_muscle.ray.intersects_location(inside_verts.squeeze().detach().cpu().numpy(), 
                                        inside_verts_normal.squeeze().detach().cpu().numpy())
        locations = locations + 2* inside_verts_normal.squeeze().detach().cpu().numpy()[index_ray] # outside 2 mm
        index_ray = torch.from_numpy(index_ray).to(self.device)
        offset = torch.zeros_like(inside_verts)
        offset[index_ray] = torch.from_numpy(locations).float().to(self.device) - inside_verts[index_ray]

        ## move
        skin_v_offset = torch.zeros_like(floating_verts[i])
        skin_v_offset[skin_surf_in_muscle] = offset

        hard_result = floating_verts[i] + skin_v_offset
        soft_result = floating_verts[i] + (self.skin_v_node_weight.unsqueeze(-1) * skin_v_offset).sum(1)
        final_result = 0.7 * hard_result + 0.3 * soft_result
        floating_verts[i] = final_result
        
    return floating_verts
```

But if I set `handle_collision` to false, the hand collision problem will occur. May I ask how to solve this problem? 

Thanks!

**Comment by reyuwei** [comment]
I only implemented this function with Trimesh, so it has to be in cpu. You can try and implement it with other gpu-based libraries, you will need to replace `mesh_muscle.contains` and `mesh_muscle.ray.intersects_location`.  

You can also use the [contact loss](https://github.com/reyuwei/nr-reg/blob/27570ca7f12a1b4f62d952f20154f2b865028bde/loss/loss_collecter.py#L365) described in our paper to handle collision.

**Comment by delaprada** [comment]
> I only implemented this function with Trimesh, so it has to be in cpu. You can try and implement it with other gpu-based libraries, you will need to replace `mesh_muscle.contains` and `mesh_muscle.ray.intersects_location`.
> 
> You can also use the [contact loss](https://github.com/reyuwei/nr-reg/blob/27570ca7f12a1b4f62d952f20154f2b865028bde/loss/loss_collecter.py#L365) described in our paper to handle collision.

Thank you very much for your reply! I will try your suggestion.

By the way, does this contact loss act as a regularization term to make the results plausible too?

**Comment by reyuwei** [comment]
> > I only implemented this function with Trimesh, so it has to be in cpu. You can try and implement it with other gpu-based libraries, you will need to replace `mesh_muscle.contains` and `mesh_muscle.ray.intersects_location`.
> > You can also use the [contact loss](https://github.com/reyuwei/nr-reg/blob/27570ca7f12a1b4f62d952f20154f2b865028bde/loss/loss_collecter.py#L365) described in our paper to handle collision.
> 
> Thank you very much for your reply! I will try your suggestion.
> 
> By the way, does this contact loss act as a regularization term to make the results plausible too?

The contact loss was designed to make sure bone and muscle is collision free. I guess it can also be viewed as a regularization term, as collision free indeed is more plausible.
